ta/trusted_key/entry.c

```diff
@@ -0,0 +1,308 @@
+		params[0].memref.size = extra_size;
+	}
+
+	params[1].memref.buffer = key;
+	params[1].memref.size = key_size;
+
+	res = TEE_InvokeTACommand(sess, TEE_TIMEOUT_INFINITE,
+				  PTA_SYSTEM_DERIVE_TA_UNIQUE_KEY,
+				  param_types, params, &ret_orig);
+
+	TEE_CloseTASession(sess);
+
+	return res;
+}
+
+static TEE_Result huk_AE_encrypt(TEE_OperationHandle crypto_op, void *in,
+				 size_t in_sz, void *out, size_t *out_sz)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	uint8_t *iv = (uint8_t *)out;
+	uint8_t *tag = (uint8_t *)out + IV_SIZE;
+	uint8_t *enc_key = (uint8_t *)out + IV_SIZE + TAG_SIZE;
+	size_t enc_key_len = in_sz, tag_len = TAG_SIZE;
+
+	TEE_GenerateRandom(iv, IV_SIZE);
+
+	res = TEE_AEInit(crypto_op, iv, IV_SIZE, TAG_SIZE * 8, 0, 0);
+	if (res != TEE_SUCCESS)
+		return res;
+
+	res = TEE_AEEncryptFinal(crypto_op, in, in_sz, enc_key, &enc_key_len,
+				 tag, &tag_len);
+	if (res != TEE_SUCCESS || tag_len != TAG_SIZE) {
+		res = TEE_ERROR_SECURITY;
+		return res;
+	}
+
+	if (ADD_OVERFLOW(enc_key_len, (IV_SIZE + TAG_SIZE), out_sz))
+		return TEE_ERROR_SECURITY;
+
+	return res;
+}
+
+static TEE_Result huk_AE_decrypt(TEE_OperationHandle crypto_op, void *in,
+				 size_t in_sz, void *out, size_t *out_sz)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	uint8_t *iv = (uint8_t *)in;
+	uint8_t *tag = (uint8_t *)in + IV_SIZE;
+	uint8_t *enc_key = (uint8_t *)in + IV_SIZE + TAG_SIZE;
+	size_t enc_key_len = 0;
+
+	if (SUB_OVERFLOW(in_sz, (IV_SIZE + TAG_SIZE), &enc_key_len))
+		return TEE_ERROR_SECURITY;
+
+	res = TEE_AEInit(crypto_op, iv, IV_SIZE, TAG_SIZE * 8, 0, 0);
+	if (res != TEE_SUCCESS)
+		return res;
+
+	res = TEE_AEDecryptFinal(crypto_op, enc_key, enc_key_len, out, out_sz,
+				 tag, TAG_SIZE);
+	if (res != TEE_SUCCESS)
+		res = TEE_ERROR_SECURITY;
+
+	return res;
+}
+
+static TEE_Result huk_crypt(TEE_OperationMode mode, void *in, size_t in_sz,
+			    void *out, size_t *out_sz)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	TEE_OperationHandle crypto_op = TEE_HANDLE_NULL;
+	TEE_ObjectHandle hkey = TEE_HANDLE_NULL;
+	uint8_t huk_key[TA_DERIVED_KEY_MAX_SIZE] = { 0 };
+	TEE_Attribute attr = { 0 };
+
+	res = TEE_AllocateOperation(&crypto_op, TEE_ALG_AES_GCM, mode,
+				    sizeof(huk_key) * 8);
+	if (res != TEE_SUCCESS)
+		return res;
+
+	res = derive_unique_key(huk_key, sizeof(huk_key), NULL, 0);
+	if (res != TEE_SUCCESS) {
+		EMSG("derive_unique_key failed: returned 0x%x\n", res);
+		goto out_op;
+	}
+
+	res = TEE_AllocateTransientObject(TEE_TYPE_AES, sizeof(huk_key) * 8,
+					  &hkey);
+	if (res != TEE_SUCCESS)
+		goto out_op;
+
+	attr.attributeID = TEE_ATTR_SECRET_VALUE;
+	attr.content.ref.buffer = huk_key;
+	attr.content.ref.length = sizeof(huk_key);
+
+	res = TEE_PopulateTransientObject(hkey, &attr, 1);
+	if (res != TEE_SUCCESS)
+		goto out_key;
+
+	res = TEE_SetOperationKey(crypto_op, hkey);
+	if (res != TEE_SUCCESS)
+		goto out_key;
+
+	if (mode == TEE_MODE_ENCRYPT) {
+		res = huk_AE_encrypt(crypto_op, in, in_sz, out, out_sz);
```

Comment:
Shouldn't some kind of version be included in the output?
The use case I have in mind is an upgrade of the TEE, where the new version would use a different algorithm by default.

Reply:
Binding a trusted key to a particular TEE version might not be a good idea as user may want to backport this early TA to older versions of TEE.

But the use case you pointed out seems to be valid. So in order to address it, I can add algorithm identifier to output trusted key blob and use that identifier during unseal operation. Does it suffice your use case?

Reply:
For maximum flexibility **and** simplicity I would just add an 8-bit field to the output (`uint8_t reserved`) and make sure it is set to zero for now. If anything needs to change later, we can make this field a version number (0, 1, 2 etc.) or whatever.

Reply:
Okay.